documentation/docs/core-concepts/basic-principles.md

```diff
@@ -0,0 +1,56 @@
+---
+title: Basic Principles
+path: /basic-principles/
+---
+
+# Basic Principles
+
+Internally, Fusion.js uses [Koa](https://koajs.com/) to handle request routing using a middleware stack. If you are unfamiliar with the concept of a middleware stack, think of an incoming request being processed by multiple layers of functions, that is, "stacking" on top of each other. As each function executes, it processes the request and passes it to the next middleware in the stack. For example, an incoming request might require some data to be fetched from a database first, then passes that information to a second middleware that uses that data to make a calculation.
+
+While traditionally middleware is more the realm of server side frameworks (see [expressjs](https://expressjs.com/en/guide/writing-middleware.html) or [connect](https://github.com/senchalabs/connect)), because of the universal nature of Fusion.js, the middleware stack is processed on **both the client and the server**.
+
+For example, a basic Fusion.js middleware function might look like this:
+
+```js
+const app = new App(rootComponent);
+
+app.middleware(
+  () => async (ctx, next) => {
+    // Do something with the request here
+    // Here, we'll just do something useless
+    console.log('Inside middleware!');
+    // Call next to go to the next middleware in the stack
+    next();
+  }
+);
+```
+
+Here, we are attaching a middleware to `FusionApp` and adding logic to the request chain. We are telling the framework — _"anytime a request comes in, run that request through the middleware function that we just defined"_.
+
+However, keep in mind this happens on the client as well. So when the page loads, a `console.log` statement will write to `stdout` on the Node server as well as to the console on your web browser. The timing and ordering of when that occurs is a bit nuanced but is covered in [Anatomy of a Request](/docs/core-concepts/anatomy-of-a-request).
+
+You can clearly see how extensible this is. You can register multiple middleware methods on a Fusion.js app to achieve different things:
```

Comment:
Most of these examples are server-side. Was that intentional? Do you want to also talk about browser-specific examples, or examples that mix browser and server? Or do you think that is better in the "Universal framework" section?

Reply:
Is there anything special that Fusion provides for client side development? From my perspective once the server hydrates down to the client, everything after that is typical React client side SPA development. Unless I'm forgetting something obvious here that we provide.

I think the more complicated bits relate to universal rendering e.g. hydration of state to be read from the client or using `prepare` to fully load a page before flushing it down to the client.

In general though I was trying to keep these sections short and sweet unless there is anything specific you wanted me to call out here.

Reply:
Yeah I was thinking about the universal rendering examples; I think it's fine to keep that in the universal rendering section.